pkg/parse/run.go

```diff
@@ -153,95 +154,57 @@ func Run(ctx context.Context, p Parser, nsControllerState *namespacecontroller.S
-				// Don't reset retryTimer if retry limit has been reached.
-				continue
+				// Tell the event producer that we did NOT call the RunFunc.
+				return events.Result{DelayStatusUpdate: false}
 			}
 
-			retryDuration := state.backoff.Step()
-			retries := retryLimit - state.backoff.Steps
-			klog.Infof("a retry is triggered (trigger type: %v, retries: %v/%v)", trigger, retries, retryLimit)
 			// During the execution of `run`, if a new commit is detected,
 			// retryTimer will be reset to `Options.RetryPeriod`, and state.backoff is reset to `defaultBackoff()`.
 			// In this case, `run` will try to sync the configs from the new commit instead of the old commit
 			// being retried.
-			runFn(ctx, p, trigger, state)
-			// Reset retryTimer after `run` to make sure `retryDuration` happens between the end of one execution
-			// of `run` and the start of the next execution.
-			retryTimer.Reset(retryDuration)
-			statusUpdateTimer.Reset(opts.StatusUpdatePeriod) // Schedule status update attempt
-
-		// Update the sync status to report management conflicts (from the remediator).
-		case <-statusUpdateTimer.C():
-			// Publish the sync status periodically to update remediator errors.
-			// Skip updates if the remediator is not running yet, paused, or watches haven't been updated yet.
-			// This implies that this reconciler has successfully parsed, rendered, validated, and synced.
-			if opts.Remediating() {
-				klog.V(3).Info("Updating sync status (periodic while not syncing)")
-				// Don't update the sync spec or commit.
-				if err := setSyncStatus(ctx, p, state, state.status.SyncStatus.Spec, false, state.status.SyncStatus.Commit, p.SyncErrors()); err != nil {
-					klog.Warningf("failed to update sync status: %v", err)
-				}
-			}
+			result := runOpts.Run(ctx, p, trigger, state)
 
-			statusUpdateTimer.Reset(opts.StatusUpdatePeriod) // Schedule status update attempt
-			// we should not reset retryTimer under this `case` since it is not aware of the
-			// state of backoff retry.
-
-		// Execute the entire parse-apply-watch loop for a namespace event.
-		case <-nsEventTimer.C():
-			if nsControllerState == nil {
-				// If the Namespace Controller is not running, stop the timer without
-				// closing the channel.
-				nsEventTimer.Stop()
-				continue
-			}
-			if nsControllerState.ScheduleSync() {
-				klog.Infof("A new sync is triggered by a Namespace event")
-				// Reset the cache partially to make sure all the steps of a parse-apply-watch loop will run.
-				// The cached sourceState will not be reset to avoid reading all the source files unnecessarily.
-				// The cached needToRetry will not be reset to avoid resetting the backoff retries.
-				state.resetPartialCache()
-				runFn(ctx, p, namespaceEvent, state)
+			// If the commit changed, then reset the retry backoff for the new commit.
+			if result.CommitChanged {
+				return events.Result{
+					DelayStatusUpdate: true,
+					ResetRetryBackoff: true,
```

Comment:
It only resets retry backoff when source changed. 
The retry backoff also needs to be reset when a run succeeded. Otherwise, the retry backoff continues from the previous failure before the success.

Reply:
Added a Success flag to the RunResult for this.